substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageBFDNameProvider.java

```diff
@@ -441,30 +627,75 @@ public String mangle(Member member) {
 
-        private void manglePrefix(String prefix) {
-            if (!substitutePrefix(prefix)) {
-                mangleSimpleName(prefix);
+        private void mangleWriteSubstitutableNameNoRecord(String name) {
+            LookupName lookupName = new SimpleLookupName(name);
+            if (!substituteName(lookupName)) {
+                // failed so mangle the name
+                mangleWriteSimpleName(name);
             }
         }
 
-        private boolean substitutePrefix(String prefix) {
-            int index = prefixIdx(prefix);
+        private void mangleWriteSubstitutablePrefixedName(String prefix, String name) {
+            // this should only be called when inserting a sequence into a namespace
+            assert sb.charAt(sb.length() - 1) == 'N';
+            // we can substitute both symbols
+            mangleWriteSubstitutableNameRecord(prefix);
+            // In theory the trailing name at the end of the namespace ought
+            // to be able to be encoded with a short name i.e. we ought to be
+            // able to call mangleWriteSubstitutableNameNoRecord(name) at this
+            // point. Indeed, the demangler *will* translate a substitution when it
+            // is inserted at the end of a namespace e.g.
+            //
+            // c++filt _ZN3foo3barS_E --> foo::bar::foo
+            //
+            // However, gdb will barf on the $_ and refuse to translate the symbol
+            // So, in order to keep gdb happy we have to avoid translating this
+            // final name ane emit it as a simple name e.g. with the above example
+            // we would generate _ZN3foo3bar3fooE.
+
+            mangleWriteSimpleName(name);
+        }
+
+        private void mangleWriteSubstitutablePrefixedName(String prefix1, String prefix2, String name) {
+            // this should only be called when inserting a sequence into a namespace
+            assert sb.charAt(sb.length() - 1) == 'N';
+            // we can substitute the composed prefix followed by the name
+            // or we can substitute all three individual symbols
+            LookupName lookup2 = new SimpleLookupName(prefix2);
+            LookupName lookup = new NamespaceLookupName(prefix1, lookup2);
+            // try substituting the combined prefix
+            if (!substituteName(lookup)) {
+                // we may still be able to establish a binding for the first prefix
+                mangleWriteSubstitutableNameRecord(prefix1);
+                // we cannot establish a binding for the trailing prefix
+                mangleWriteSubstitutableNameNoRecord(prefix2);
+                recordName(lookup);
+            }
+            // see comment in previous method as to why we call mangleWriteSimpleName
+            // instead of mangleWriteSubstitutableNameNoRecord(name)
+            mangleWriteSimpleName(name);
+        }
+
+        private boolean substituteName(LookupName name) {
+            int index = bindings.indexOf(name);
```

Comment:
Linear search ... are we sure that this does not explode (computationally)

Reply:
The lookup cost here will be o(N^2) where N is the number of substitutable elements in the name and signature. N accumulates at most

  1 for the package (namespace) prefix
  2 for an object return type
  2 for an object or (any) array type argument
  1 if we have a boolean argument
  1 if we have a char argument
  1 if we have a byte argument

The count for an object argument/return type can be lowered if

  the object type is a foreign type (not passed as a pointer) - subtract 1
  the object type repeats the namespace prefix (e.g. java.lang.String) - subtract 1
  the object type repeats a previous occurrence - subtract 2

I added a temporary tweak to the patch to count the bindings length for each symbol encoded and tested it by building the image used for the debuginfo test. The worst case I found was this one:
 
```
java.util.Collection*
sun.security.provider.certpath.DistributionPointFetcher::getCRLs(java.security.cert.X509CRLSelector*,
sun.security.x509.X509CertImpl*, sun.security.x509.DistributionPoint*, boolean[]*,
boolean, java.security.PublicKey*, java.security.cert.X509Certificate*, java.lang.String*,
java.util.List*, java.util.Set*, java.util.Date*, java.lang.String*, java.security.cert.TrustAnchor*)
```
Encoding it generates 26 bindings. It has a lot of args and the only repeat is the second occurrence of `java.lang.String*` which appears in the encoding as `SG_` i.e. binding 18.

That case is very much an outlier. Normally the number of bindings is quite small. Here is a table summarizing the distribution for the 65667 different encodings generated during generation of one of the test images:

|Bindings Count | Frequency  | %ge |
|-:|-:|-:|
|1 | 12721| 19.4|
|2 | 7482| 11.4|
|3 | 24745| 37.7|
|4 | 6049| 9.2|
|5 | 8164| 12.4|
|6 | 2509| 3.8|
|7 | 2578| 3.9|
|8 | 358| 5.4|
|9 | 665| 1.0|
|10 | 108| 0.1|
|11-15 | 260|0.39|
|16-20 | 24| ~0.0|
|20-25 | 3| ~0.0|
|26 | 1| ~0.0|

I don't think this means we must switch to a constant time lookup model such as an EconomicMap but the trade off is not clear. What do you think?

Reply:
Thanks for the detailed breakdown.

Of course one can think of pathological methods (potentially from generated source code or test suites) that could be much worse than what you found. The question is if it's practically relevant. I would suggest you build a large test image (huge quarkus or springboot native app ... @zakkak might be able to help with that) and check if you can see any real world build time regressions that are distinguishable from noise when building with `-g`.

Reply:
I could even use the "temporary tweak to the patch to count the bindings length" to report some more stats, other than the build times. @adinn would you mind sharing that patch, to make sure we are reporting the same thing?

Reply:
@zakkak Not much to share really. The temporary tweak was simply a call to `System.out.format()` at the end of `BFDMangle::mangle(String, ResolvedJavaType, String, Signature, boolean)`, printing the resulting mangled name and the length of `bindings`. The rest of the work was done using shell commands on the resulting output.

Reply:
@adinn this is a kind reminder

Reply:
Well it looks like I was looking at a cached version of this when I pinged you :/

Thanks for the reply @adinn , I will try to get some number (hopefully this week)